### ISSUE TITLE
When loading via a promise, bad error: "  Error: d is undefined " is given

### ISSUE THREAD

**Issue by StuAtGit** [Siyfion/angular-typeahead#77]
I pull my data out via a promise, and set up the typeahead in the promises callback like so:

```
var callback = function( data ) {
		var brandsBloodhound = new Bloodhound({
			datumTokenizer : function(datum) { return Bloodhound.tokenizers.whitespace(datum.name);},
			queryTokenizer : Bloodhound.tokenizers.whitespace,
			identify: function(datum) { return datum.name; },
			local : data,
			initialize : true
		});
		$scope.typeaheadData = {};

		$scope.typeaheadData.options = {
			highlight: true
		};
		$scope.typeaheadData.brandDataset = {
			displayKey : "name",
			source : brandsBloodhound.ttAdapter()
		};
		$scope.typeaheadData.selected = data[0].name;
	};

	$myService.getData().then(
		function( data ) {
			loadBrands( data );
		},
		function( error ) {
			$scope.errorDisplay = error;
		}
	);
```

And you end up getting an error like:

```
18:25:48.842 Error: d is undefined
attach/<@http://localhost:63342/video-reviewer-fe/bower_components/typeahead.js/dist/typeahead.bundle.js:2240:40
reverseArgs@http://localhost:63342/video-reviewer-fe/bower_components/typeahead.js/dist/typeahead.bundle.js:968:28
.each@http://localhost:63342/video-reviewer-fe/bower_components/jquery/dist/jquery.js:374:14
_</<.each@http://localhost:63342/video-reviewer-fe/bower_components/typeahead.js/dist/typeahead.bundle.js:966:17
attach@http://localhost:63342/video-reviewer-fe/bower_components/typeahead.js/dist/typeahead.bundle.js:2239:21
.each@http://localhost:63342/video-reviewer-fe/bower_components/jquery/dist/jquery.js:374:14
jQuery.prototype.each@http://localhost:63342/video-reviewer-fe/bower_components/jquery/dist/jquery.js:139:10
initialize@http://localhost:63342/video-reviewer-fe/bower_components/typeahead.js/dist/typeahead.bundle.js:2236:1
$.fn.typeahead@http://localhost:63342/video-reviewer-fe/bower_components/typeahead.js/dist/typeahead.bundle.js:2383:24
initialize@http://localhost:63342/video-reviewer-fe/bower_components/angular-typeahead/angular-typeahead.js:102:13
.link@http://localhost:63342/video-reviewer-fe/bower_components/angular-typeahead/angular-typeahead.js:17:9
cloneAndAnnotateFn/<@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:8774:41
invokeLinkFn@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:8780:9
nodeLinkFn@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:8280:1
compositeLinkFn@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:7671:13
compositeLinkFn@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:7675:13
compositeLinkFn@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:7675:13
compositeLinkFn@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:7675:13
publicLinkFn@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:7546:30
ngViewFillContentFactory/<.link@http://localhost:63342/video-reviewer-fe/bower_components/angular-route/angular-route.js:985:7
invokeLinkFn@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:8780:9
nodeLinkFn@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:8280:1
compositeLinkFn@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:7671:13
publicLinkFn@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:7546:30
createBoundTranscludeFn/boundTranscludeFn@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:7690:1
controllersBoundTransclude@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:8307:18
update@http://localhost:63342/video-reviewer-fe/bower_components/angular-route/angular-route.js:935:25
$RootScopeProvider/this.$get</Scope.prototype.$broadcast@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:16244:15
commitRoute/<@http://localhost:63342/video-reviewer-fe/bower_components/angular-route/angular-route.js:619:15
processQueue@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:14678:28
scheduleProcessQueue/<@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:14694:27
$RootScopeProvider/this.$get</Scope.prototype.$eval@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:15922:16
$RootScopeProvider/this.$get</Scope.prototype.$digest@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:15733:15
$RootScopeProvider/this.$get</Scope.prototype.$apply@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:16030:13
done@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:10545:36
completeRequest@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:10717:7
requestLoaded@http://localhost:63342/video-reviewer-fe/bower_components/angular/angular.js:10658:1
 <input class="typeahead ng-pristine ng-untouched ng-valid ng-isolate-scope" sf-typeahead="" options="typeaheadData.options" datasets="typeaheadData.brandDataset" ng-model="typeaheadData.selected" type="text">1 angular.js:12450:18
consoleLog/<() angular.js:12450
$ExceptionHandlerProvider/this.$get</<() angular.js:9237
invokeLinkFn() angular.js:8782
nodeLinkFn() angular.js:8280
compositeLinkFn() angular.js:7671
compositeLinkFn() angular.js:7675
compositeLinkFn() angular.js:7675
compositeLinkFn() angular.js:7675
publicLinkFn() angular.js:7546
ngViewFillContentFactory/<.link() angular-route.js:985
invokeLinkFn() angular.js:8780
nodeLinkFn() angular.js:8280
compositeLinkFn() angular.js:7671
publicLinkFn() angular.js:7546
createBoundTranscludeFn/boundTranscludeFn() angular.js:7690
controllersBoundTransclude() angular.js:8307
update() angular-route.js:935
$RootScopeProvider/this.$get</Scope.prototype.$broadcast() angular.js:16244
commitRoute/<() angular-route.js:619
processQueue() angular.js:14678
scheduleProcessQueue/<() angular.js:14694
$RootScopeProvider/this.$get</Scope.prototype.$eval() angular.js:15922
$RootScopeProvider/this.$get</Scope.prototype.$digest() angular.js:15733
$RootScopeProvider/this.$get</Scope.prototype.$apply() angular.js:16030
done() angular.js:10545
completeRequest() angular.js:10717
requestLoaded() angular.js:10658

```

**Comment by StuAtGit** [comment]
This error is caused because typehead tries to initialize, even with the variables in it's directive have not been initialized. I worked around this by adding an "ng-if" to my tag:

So, change this:
```
            <input
                   class="typeahead"
                   type="text"
                   sf-typeahead
                   options="typeaheadData.options"
                   datasets="typeaheadData.dataset"
                   ng-model="typeaheadData.selected">
```
to:
```
<input 
    >>>>>>   ng-if="typeaheadData.brandDataset" <<<<<<<<<
                   class="typeahead" 
                   type="text"
                   sf-typeahead
                   options="typeaheadData.options"
                   datasets="typeaheadData.brandDataset"
                   ng-model="typeaheadData.selected">
```

**Comment by StuAtGit** [comment]
The best solution would be to not have the directive try to load undefined variables, and just wait until the data is available. But, not as good, but simpler would be to update this code:

```
initialize: function initialize(o, datasets) {
                var www;
                datasets = _.isArray(datasets) ? datasets : [].slice.call(arguments, 1);
```
To log some sort of clear error, for example:
```
initialize: function initialize(o, datasets) {
                if( typeof datasets === "undefined" ) {
                    console.log("Warning: no data set defined - perhaps you're trying to load your dataset asynchronous? Use ng-if to disable the typahead directive until the dataset is available if so");
                }
                var www;
                datasets = _.isArray(datasets) ? datasets : [].slice.call(arguments, 1);
```

**Comment by RongBranovate** [comment]
@StuAtGit , I've implemented your solution with the `ng-if` but it seems that the dataset is never available because my input never appears...

this is how my code looks: 
**on the controller:**
```javascript
var callback = function (data) {
        // Instantiate the bloodhound suggestion engine
        var myTitles = new Bloodhound({
            datumTokenizer: function (d) {
                return Bloodhound.tokenizers.whitespace(d.title);
            },
            queryTokenizer: Bloodhound.tokenizers.whitespace,
            local: data
        });

        // initialize the bloodhound suggestion engine
        myTitles.initialize();

        // Allows the addition of local datum
        // values to a pre-existing bloodhound engine.
        function addTitles(titlesToAdd) {
            myTitles.add(titlesToAdd);
        };

        // Typeahead options object
        $scope.myOptions = {
            highlight: true
        };

        // Single dataset example
        $scope.myData = {
            displayKey: 'title',
            source: myTitles.ttAdapter()
        };
    }

getData().then(function (response) {
            callback(response.data);
        });
```

**and on the html:**
```html
<input type="text"
                   ng-if="myData"
                   class="sfTypeahead" options="myOptions" datasets="myData"
                   ng-model="searchQuery"
                   ng-model-options="{debounce:500}">
```

what am i doing wrong?

**Comment by StuAtGit** [comment]
Hello Rod,
  So I just checked my email and saw this message - it doesn't show up on
github anymore, so maybe you've fixed it and removed the message?
Honestly, it's been a while since I looked at this, but you're solution
with the ng-if looks equivalent to mine.

I did note one difference. You have:
 > class="sf-typeahead"
But in my example, I appeared to use it as an attribute, and the class was
just "typeahead", i.e.

>ng-if="myData"
>sf-typeahead
>class="typeahead"

So I think sf-typeahead is just an attribute? But honestly, not sure, it's
been a while.

If you're still seeing a problem, what is the error you're getting? That
might help me narrow it down.

Take are,
  -stu
